helix-term/src/commands/typed.rs

```diff
@@ -71,8 +76,10 @@ fn buffer_close_by_ids_impl(
     doc_ids: &[DocumentId],
     force: bool,
 ) -> anyhow::Result<()> {
-    for &doc_id in doc_ids {
-        editor.close_document(doc_id, force)?;
+    log::debug!("closing buffers: {:?}", doc_ids);
+
+    for doc_id in doc_ids {
+        tokio::task::block_in_place(|| helix_lsp::block_on(editor.close_document(*doc_id, force)))?;
```

Comment:
Why the double `block_on` here? We should avoid blocking if possible since I've managed to get deadlocks with `block_on` + `I/O` before

Reply:
See the [docs](https://docs.rs/tokio/1.20.1/tokio/task/fn.block_in_place.html)

>Runs the provided blocking function on the current thread without blocking the executor.
>
>In general, issuing a blocking call or performing a lot of compute in a future without yielding is problematic, as it may prevent the executor from driving other tasks forward. Calling this function informs the executor that the currently executing task is about to block the thread, so the executor is able to hand off any other tasks it has to a new worker thread before that happens. See the CPU-bound tasks and blocking code section for more information.

So the outer `block_in_place` does not block, but just runs the provided blocking function and tells the executor to move other tasks off this thread, specifically to avoid deadlocking, like you said. I actually ran into this myself, which is how I found this helper.

Avoiding blocking will be difficult to say the least. By the end of the command, the buffer needs to be closed, so we have to make sure all the writes on the document are done. This is similar to shutting down the application: you have to make sure all the writes are done before quitting, like we do now by waiting on the jobs before quitting. You *must* block at some point or another.